Generate manpages

alban opened the pull request:
manpages: generate manpages with cobra

To update the manpages, use:
```
    $ make manpages
```

It will generate the manpages in dist/manpages/.

Also the bash completion file is no longer built manually, but you can
use:
```
    $ make bash-completion
```

It will generate dist/bash_completion/rkt.bash.

The release guide (release.md) is updated to make sure that each release
contains an up-to-date version of the manpages and the bash completion
file.

Fixes #2370

iaguis commented on rkt/rkt.go:
```diff
@@ -223,10 +224,23 @@ func main() {
+		os.Exit(0)
+	}
+
+	// update rkt.bash
```
wrong comment

iaguis commented on rkt/rkt.go:
```diff
@@ -223,10 +224,23 @@ func main() {
+			Section: "1",
+		}
+
+		doc.GenManTree(cmdRkt, header, "dist/manpages")
```
Maybe we should not ignore errors?

alban replied:
I will fix.

alban commented:
Assigning to @s-urbaniak to get a review.

s-urbaniak commented:
Seems odd that we are using the `rkt` binary to generate man pages by setting a magic env var and also checking in generated files. Generation of man pages is something we should rather do at build time only and not pollute the runtime behavior of `rkt`?!

How about having a separate `man.go` file having a `+man` build flag being invoked as a make target `make doc` to generate the man pages explicitly using `go run`?

s-urbaniak commented on rkt/rkt.go:
```diff
@@ -223,10 +224,23 @@ func main() {
-	// // Uncomment to update rkt.bash
-	// stdout.Print("Generating rkt.bash")
-	// cmdRkt.GenBashCompletionFile("dist/bash_completion/rkt.bash")
-	// os.Exit(0)
```
:+1: for removing commented code

jellonek commented on dist/manpages/rkt_fetch.1:
```diff
@@ -0,0 +1,78 @@
+
+.PP
+\fB\-\-insecure\-options\fP=none
+    comma\-separated list of security features to disable. Allowed values: "none", "image", "tls", "ondisk", "http", "pubkey", "all"
```
IMO at least this part deserves for wider description.

alban replied:
This would be for another pull request.

alban commented:
@s-urbaniak PR updated. I also updated the description above. PTAL.

s-urbaniak commented on rkt/manpages_gen.go:
```diff
@@ -0,0 +1,39 @@
+
+	err := doc.GenManTree(cmdRkt, header, "dist/manpages")
+	if err != nil {
+		fmt.Printf("Failed to generate the man pages: %v\n", err)
```
nit: use stderr instead?

alban replied:
ok

s-urbaniak commented on rkt/main.go:
```diff
@@ -0,0 +1,36 @@
+	// Make help just show the usage
+	cmdRkt.SetHelpTemplate(`{{.UsageString}}`)
+
+	if err := cmdRkt.Execute(); err != nil && cmdExitCode == 0 {
```
nit: we ignore the `err`? I know this is copied, but it seems, the err is swallowed.

alban replied:
It seems already printed by Cobra... I tried to add:
```
		fmt.Fprintf(os.Stderr, "Execute failed: %v\n", err)
```

But then, the error appears twice:

```
$ ./build-rkt-1.3.0+git/bin/rkt gggfdgfd
Error: unknown command "gggfdgfd" for "rkt"
Run 'rkt --help' for usage.
Execute failed: unknown command "gggfdgfd" for "rkt"
```

s-urbaniak replied:
Interesting, thanks for pointing out. Let's maybe leave a `// err printed by cobra on stderr` then?

alban replied:
comment added below.

s-urbaniak commented on rkt/bash-completion_gen.go:
```diff
@@ -0,0 +1,32 @@
+func main() {
+	err := cmdRkt.GenBashCompletionFile("dist/bash_completion/rkt.bash")
+	if err != nil {
+		fmt.Printf("Failed to generate the bash completion file: %v\n", err)
```
nit: use stderr?

alban replied:
fixed

s-urbaniak commented:
@alban small nits, I am still unsure of checking in the generated man files (@jonboulle, comments?), else LGTM :-)

jonboulle commented:
I agree with @s-urbaniak that we should avoid committing the generated artefacts - otherwise we're inevitably going to have rot between them. Can't we just keep the generation as part of the release process and in notes for distributors?

alban commented:
@s-urbaniak @jonboulle Updated. I added new commits and I will squash them once it looks good to you. I removed the generated files from the git repository (except the protobuf generated files, see #2004 - to be done in a different PR), and updated the build scripts.

s-urbaniak commented on CHANGELOG.md:
```diff
@@ -1,3 +1,12 @@
+## vUNRELEASED
```
the `v` is superfluous.

krnowak replied:
We usually prepend `v` to version number, so I think it's ok.

s-urbaniak replied:
got it, thanks!

s-urbaniak commented:
LGTM

krnowak commented on scripts/build-rkt.sh:
```diff
@@ -2,4 +2,11 @@
 set -e
 
-./autogen.sh && ./configure --enable-tpm=no --with-stage1-default-images-directory=/usr/lib/rkt/stage1-images --with-stage1-default-location=/usr/lib/rkt/stage1-images/stage1-coreos.aci && make -j4
+./autogen.sh && \
```
Hm, we have the `set -e` call above so chaining everything with `&&` is not really necessary. Could you simplify it by removing those `&&` and removing one level of indentation in the lines below?

alban replied:
ok

krnowak commented on rkt/rkt.mk:
```diff
@@ -32,3 +32,11 @@ include makelib/build_go_bin.mk
+
+bash-completion:
+	mkdir -p dist/bash_completion/
+	ls rkt/*.go | grep -vE '_test.go|main.go|_gen.go' | xargs go run rkt/bash-completion_gen.go
```
bash**dash**competion**underline**gen.go? Ew!

alban replied:
Should I use bash_completion_gen.go instead?

krnowak commented on dist/bash_completion/rkt.bash:
```diff
@@ -1,860 +0,0 @@
-#!/bin/bash
```
And I remember manually editing this file when removing the `--insecure-skip-verify` flag, stupid me.

krnowak commented on rkt/rkt.mk:
```diff
@@ -32,3 +32,11 @@ include makelib/build_go_bin.mk
 # RKT_STAMP deliberately not cleared
 # RKT_BINARY deliberately not cleared
+
+manpages:
```
/me turns a blind eye on this. The concerns I have, and which maybe I'll address later when I simplify some bits and document them, are:
- I would make it a part of `make all`.
- I would prefer the files to be generated inside `$(BUILDDIR)`
  - That would require passing that directory to the generator, so it would know where to put the files.
  - But that complicates things (make clean, deps, forwarding variables, blah blah blah), so don't bother with it for now.
- Pretty printers could be used there.
  - No big deal, since the output is rather short.
- These additions should be put above the `$(call undefine-namespaces,LOCAL)` line.

alban replied:
> I would make it a part of make all.

It would make the compilation quite slow if we run `make manpages/bash-completion` at each `make` though... On my laptop:

- time make manpages: 25 seconds
- time make bash-completion: 24 seconds
- total: 49 seconds

> I would prefer the files to be generated inside $(BUILDDIR)
> That would require passing that directory to the generator, so it would know where to put the files.
> But that complicates things (make clean, deps, forwarding variables, blah blah blah), so don't bother with it for now.

Ok, I'll skip that for now.

> Pretty printers could be used there.
> No big deal, since the output is rather short.

How to do that?

> These additions should be put above the $(call undefine-namespaces,LOCAL) line.

Will do in the next update.

krnowak replied:
> It would make the compilation quite slow if we run `make manpages/bash-completion` at each `make` though... On my laptop:
>
> - time make manpages: 25 seconds
- time make bash-completion: 24 seconds
- total: 49 seconds

I wanted it to be a part of `make all`, because it is about building everything. Possibly this could be controlled with `--enable-man` and `--enable-bash-completion` flags which could be disabled by default.

> > Pretty printers could be used there.
> > No big deal, since the output is rather short.
>
> How to do that?

```
manpages:
	$(VQ) \ <- this expands to @ when pretty printing (V=0,1,2) or to nothing for raw output (V=3)
	set -e;
	$(call vb,vt,GEN,manpages) \ <- this will print (GEN     manpages for V=1,2, nothing otherwise)
	ls rkt/*.go | blah blah blah
```

alban commented:
Updated. PTAL.

iaguis commented:
Running

```
$ ./autogen.sh && ./configure --enable-tpm=no && make manpages
```

doesn't work for me because before make, there's nothing in `$BUILDDIR/gopath`:

```
mkdir -p dist/manpages/
ls rkt/*.go | grep -vE '_test.go|main.go|_gen.go' | xargs go run rkt/manpages_gen.go
rkt/image_cmd_common.go:24:2: cannot find package "github.com/appc/spec/discovery" in any of:
	/usr/lib/go/src/github.com/appc/spec/discovery (from $GOROOT)
	/home/iaguis/work/coreos/go/src/github.com/coreos/rkt/build-rkt-1.3.0+git/gopath/src/github.com/coreos/rkt/Godeps/_workspace/src/github.com/appc/spec/discovery (from $GOPATH)
	/home/iaguis/work/coreos/go/src/github.com/coreos/rkt/build-rkt-1.3.0+git/gopath/src/github.com/appc/spec/discovery
rkt/api_service.go:28:2: cannot find package "github.com/appc/spec/schema" in any of:
	/usr/lib/go/src/github.com/appc/spec/schema (from $GOROOT)
	/home/iaguis/work/coreos/go/src/github.com/coreos/rkt/build-rkt-1.3.0+git/gopath/src/github.com/coreos/rkt/Godeps/_workspace/src/github.com/appc/spec/schema (from $GOPATH)
	/home/iaguis/work/coreos/go/src/github.com/coreos/rkt/build-rkt-1.3.0+git/gopath/src/github.com/appc/spec/schema

[...]
```

krnowak commented:
Meh, as a workaround for this issue (`./autogen.sh && ./configure && make manpages`) just do:
```
manpages: | $(GOPATH_TO_CREATE)/src/$(REPO_PATH)
bash-completion: | $(GOPATH_TO_CREATE)/src/$(REPO_PATH)
```

iaguis commented:
The workaround works :construction_worker:

krnowak commented:
Cool.

iaguis commented:
LGTM

alban commented:
Ok, commits squashed.

iaguis commented:
LGTMTMOG

iaguis commented:
Nit: actually you don't update the release guide but update the release script (commit message).
